File: HorizontalPicker/HPCollectionViewFlowLayout.m
//  The MIT License (MIT)
//  Copyright (c) 2013 <RABE_IT Services>
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"),
//  to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
//  and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
//  IN THE SOFTWARE.
//
//
//
//  HPCollectionViewFlowLayout.m
//  HorizontalPicker
//

#import "HPCollectionViewFlowLayout.h"
#import "HPickerDefinitions.h"
#import "HorizontalPickerGlobal.h"

@implementation HPCollectionViewFlowLayout

- (void)prepareLayout
{
    self.itemSize                = CGSizeMake(self.itemSize.width, PICKER_HEIGHT);
    self.minimumInteritemSpacing = 0.0;
    self.minimumLineSpacing      = 0.0;
    self.scrollDirection         = UICollectionViewScrollDirectionHorizontal;
    self.sectionInset            = UIEdgeInsetsMake(0.0, 0.0, 0.0, 0.0);
}

- (NSArray *)layoutAttributesForElementsInRect:(CGRect)rect
{
    NSArray *array = [super layoutAttributesForElementsInRect:rect];
    
    if (_style == HPStyle_iOS7) {
        CGRect visibleRect = CGRectZero;
        visibleRect.origin = self.collectionView.contentOffset;
        visibleRect.size = self.collectionView.bounds.size;
        
        CGFloat activeDistance = floorf(CGRectGetWidth(visibleRect) / 2);
        CGFloat midX = ceilf(CGRectGetMidX(visibleRect));
        CGFloat baseAngle = M_PI/180 * 40;
        
        for (UICollectionViewLayoutAttributes *attributes in array)
        {
            CGFloat distance = midX - CGRectGetMidX(attributes.frame);
            attributes.alpha = (activeDistance - ABS(distance))/activeDistance;
            
            CGFloat value = -distance/activeDistance;
            value *= baseAngle;
            attributes.transform3D = CATransform3DMakeRotation(value, 0, 1, 0);
            
            attributes.zIndex = ABS(distance)/activeDistance * 200;
            
            if (DEBUG_HP == 1) {
                NSLog(@"%d (%@) alpha = %2.2f, zIndex = %d", attributes.indexPath.row, NSStringFromCGRect(attributes.frame), attributes.alpha, attributes.zIndex);
            }
        }
    }
    return array;
}

- (BOOL)shouldInvalidateLayoutForBoundsChange:(CGRect)newBounds
{
    return YES;
}


@end

File: HorizontalPicker/Classes/CollectionView/HPCollectionViewFlowlayout.swift
//
//  HPCollectionViewFlowlayout.swift
//  HorizontalPickerDemo
//
//  Created by Bernd Rabe on 13.12.15.
//  Copyright © 2015 RABE_IT Services. All rights reserved.
//

import UIKit
import GLKit

class HPCollectionViewFlowlayout: UICollectionViewFlowLayout {
    
    var activeDistance: CGFloat = 0.0
    var midX: CGFloat = 0
    var lastElementIndex = 0
    let maxAngle = CGFloat(GLKMathDegreesToRadians(HorizontalPickerViewConstants.maxRotationAngle))
    
    override func prepare () {
        minimumInteritemSpacing = 0.0
        scrollDirection = .horizontal
    }
    
    override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
        guard let array = super.layoutAttributesForElements(in: rect) , let cv = collectionView else {
            return nil
        }
        
        let visibleRect = CGRect(origin: cv.contentOffset, size: cv.bounds.size)
        var attributesCopy = [UICollectionViewLayoutAttributes]()
        
        for itemAttributes in array {
            let itemAttributesCopy = itemAttributes.copy() as! UICollectionViewLayoutAttributes
            let distance        = visibleRect.midX - itemAttributesCopy.center.x
            let normalizedDistance = distance / activeDistance
            if abs(distance) < activeDistance {
                itemAttributesCopy.alpha = 1.0 - abs(normalizedDistance)
                let rotationAngle = maxAngle * normalizedDistance
                itemAttributesCopy.transform3D = CATransform3DMakeRotation(rotationAngle, 0, 1, 0)
            } else {
                itemAttributesCopy.transform3D = CATransform3DMakeRotation(maxAngle, 0, 1, 0)
                itemAttributesCopy.alpha = 0.1
            }
            attributesCopy.append(itemAttributesCopy)
        }
        
        return attributesCopy
    }
    
    override func shouldInvalidateLayout(forBoundsChange newBounds: CGRect) -> Bool {
        return true
    }

}
